src/big/json.cr

```diff
@@ -1,44 +1,104 @@
 require "json"
 require "big"
 
-def BigInt.new(pull : JSON::PullParser)
-  pull.read_int
-  BigInt.new(pull.raw_value)
-end
+class JSON::Builder
+  # Writes a big integer.
+  def number(number : BigInt)
+    scalar do
+      @io << number
+    end
+  end
 
-def BigInt.from_json_object_key?(key : String)
-  BigInt.new(key)
-rescue ArgumentError
-  nil
-end
+  # Writes a big float.
+  def number(number : BigFloat)
+    scalar do
+      @io << number
+    end
+  end
 
-def BigFloat.new(pull : JSON::PullParser)
-  pull.read_float
-  BigFloat.new(pull.raw_value)
+  # Writes a big decimal.
+  def number(number : BigDecimal)
+    scalar do
+      @io << number
+    end
+  end
 end
 
-def BigFloat.from_json_object_key?(key : String)
-  BigFloat.new(key)
-rescue ArgumentError
-  nil
+struct BigInt
+  def self.new(pull : JSON::PullParser)
+    pull.read_int
+    new(pull.raw_value)
+  end
+
+  def self.from_json_object_key?(key : String)
+    new(key)
+  rescue ArgumentError
+    nil
+  end
+
+  def to_json_object_key
+    to_s
+  end
+
+  def to_json(json : JSON::Builder)
+    json.number(self)
+  end
 end
 
-def BigDecimal.new(pull : JSON::PullParser)
-  case pull.kind
-  when .int?
-    pull.read_int
-    value = pull.raw_value
-  when .float?
-    pull.read_float
-    value = pull.raw_value
-  else
-    value = pull.read_string
-  end
-  BigDecimal.new(value)
+struct BigFloat
+  def self.new(pull : JSON::PullParser)
+    case pull.kind
+    when .int?
+      pull.read_int
+      value = pull.raw_value
+    else
```

Comment:
Shouldn't this be in a `when .float?` branch?

Reply:
There are only two cases, either `.int?` or `.float?`, if it is not the former then it is the latter, if the else block is executed but the value is not `.float?` then this error would be thrown https://github.com/crystal-lang/crystal/blob/b1705de48d388e6e4b89dcf0ed6d5d5d54e7241d/src/json/pull_parser.cr#L222

If we do `when .float?` here, the compiler would interpret it as not having an `else` case, hence making the return value nilable, of course we can also do `else; raise ArgumentError.new("BigFloat cannot parse from invalid value: #{pull.read_raw}")` but I feel like that is unnecessarily verbose compared to the current approach.

Reply:
Hm, this might need refinement. `BigDecimal` also has a string branch. And what if it's a different kind (for example bool)?
Would be great to get some more specs for different inputs and errors.

Reply:
Sorry for this shortcoming, I'll get back on this with necessary adjustments soon.